### ISSUE TITLE
PageSelectionResolver does not resolve NODE_TYPE_INTERNAL_LINK and NODE_TYPE_EXTERNAL_LINK urls

### ISSUE THREAD

**Issue by alexandersch** [sulu/SuluHeadlessBundle#62]
The pageSelectionResolver does not resolve the correct URL when the structure's node type is NODE_TYPE_INTERNAL_LINK or NODE_TYPE_EXTERNAL_LINK. For me it just returns the internal path in the JSON response.

This is the default output right now:
```
"link": {
    "id": "6491f713-e30d-48b5-8e25-fe19e09d3582",
    "type": "page",
    "template": "default",
    "content": {
        "title": "GCTV",
        "url": "/gctv"
    },
    "view": {
        "title": [],
        "url": []
    },
    "author": "2",
    "authored": "2020-09-04T13:42:12+0200",
    "changer": 2,
    "changed": "2020-09-04T14:26:09+0200",
    "creator": 9,
    "created": "2020-07-13T13:42:12+0200"
}
```

What I would like to see is that url resolves the redirect URL like this:
```
"link": {
    "id": "6491f713-e30d-48b5-8e25-fe19e09d3582",
    "type": "page",
    "template": "default",
    "content": {
        "title": "GCTV",
        "url": "https://www.gctv.com"
    },
    "view": {
        "title": [],
        "url": []
    },
    "author": "2",
    "authored": "2020-09-04T13:42:12+0200",
    "changer": 2,
    "changed": "2020-09-04T14:26:09+0200",
    "creator": 9,
    "created": "2020-07-13T13:42:12+0200"
}
```

Example template:
```
<?xml version="1.0" ?>
<template xmlns="http://schemas.sulu.io/template/template"
          xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
          xsi:schemaLocation="http://schemas.sulu.io/template/template http://schemas.sulu.io/template/template-1.0.xsd">

    <key>test</key>

    <view>pages/default</view>
    <controller>Sulu\Bundle\HeadlessBundle\Controller\HeadlessWebsiteController::indexAction</controller>

    <meta>
        <title lang="en">Test Page</title>
    </meta>

    <properties>

        <property name="title" type="text_line" mandatory="true">
            <meta>
                <title lang="en">Title</title>
            </meta>
            <params>
                <param name="headline" value="true"/>
            </params>

            <tag name="sulu.rlp.part"/>
        </property>

        <property name="url" type="resource_locator" mandatory="true">
            <meta>
                <title lang="en">Slug</title>
            </meta>
            <tag name="sulu.rlp"/>
        </property>

        <property name="link" type="single_page_selection">
            <meta>
                <title lang="en">Link</title>
            </meta>
        </property>

    </properties>
</template>
```

**Comment by alexandersch** [comment]
I've added some code to the `PageSelectionResolver` but I think it's pretty ugly and can probably be solved in a better way.

```
--- vendor/sulu/headless-bundle/Content/ContentTypeResolver/PageSelectionResolver.php	2020-11-25 18:55:35.000000000 +0100
+++ vendor/sulu/headless-bundle/Content/ContentTypeResolver/PageSelectionResolver.php.patch	2020-11-30 14:50:04.000000000 +0100
@@ -13,15 +13,21 @@
 
 namespace Sulu\Bundle\HeadlessBundle\Content\ContentTypeResolver;
 
+use Sulu\Bundle\HeadlessBundle\Content\ContentTypeResolver\ContentTypeResolverInterface;
 use Sulu\Bundle\HeadlessBundle\Content\ContentView;
 use Sulu\Bundle\HeadlessBundle\Content\StructureResolverInterface;
+use Sulu\Bundle\PageBundle\Document\PageDocument;
 use Sulu\Component\Content\Compat\PropertyInterface;
 use Sulu\Component\Content\Compat\PropertyParameter;
+use Sulu\Component\Content\Compat\Structure;
 use Sulu\Component\Content\Mapper\ContentMapperInterface;
 use Sulu\Component\Content\Query\ContentQueryBuilderInterface;
+use Sulu\Component\DocumentManager\DocumentManagerInterface;
+use Sulu\Component\DocumentManager\Exception\DocumentManagerException;
 
 class PageSelectionResolver implements ContentTypeResolverInterface
 {
+
     public static function getContentType(): string
     {
         return 'page_selection';
@@ -43,6 +49,11 @@
     private $contentMapper;
 
     /**
+     * @var DocumentManagerInterface
+     */
+    private $documentManager;
+
+    /**
      * @var bool
      */
     private $showDrafts;
@@ -51,18 +62,29 @@
         StructureResolverInterface $structureResolver,
         ContentQueryBuilderInterface $contentQueryBuilder,
         ContentMapperInterface $contentMapper,
+        DocumentManagerInterface $documentManager,
         bool $showDrafts
-    ) {
+    )
+    {
         $this->structureResolver = $structureResolver;
         $this->contentQueryBuilder = $contentQueryBuilder;
         $this->contentMapper = $contentMapper;
+        $this->documentManager = $documentManager;
         $this->showDrafts = $showDrafts;
     }
 
+    /**
+     * @param mixed $data
+     * @param PropertyInterface $property
+     * @param string $locale
+     * @param array $attributes
+     * @return ContentView
+     * @throws DocumentManagerException
+     */
     public function resolve($data, PropertyInterface $property, string $locale, array $attributes = []): ContentView
     {
         if (empty($data)) {
-            return new ContentView([], ['ids' => []]);
+            return new ContentView([]);
         }
 
         /** @var PropertyParameter[] $params */
@@ -73,8 +95,9 @@
         $this->contentQueryBuilder->init([
             'ids' => $data,
             'properties' => $propertiesParamValue,
-            'published' => !$this->showDrafts,
+            'published' => false,
         ]);
+
         list($pagesQuery) = $this->contentQueryBuilder->build($property->getStructure()->getWebspaceKey(), [$locale]);
 
         $pageStructures = $this->contentMapper->loadBySql2(
@@ -87,7 +110,6 @@
             'title' => 'title',
             'url' => 'url',
         ];
-
         foreach ($propertiesParamValue as $propertiesParamEntry) {
             /** @var string $propertyValue */
             $propertyValue = $propertiesParamEntry->getValue();
@@ -96,9 +118,25 @@
 
         $pages = [];
         foreach ($pageStructures as $pageStructure) {
-            $pages[] = $this->structureResolver->resolveProperties($pageStructure, $propertyMap, $locale);
+
+            $page = $this->structureResolver->resolveProperties($pageStructure, $propertyMap, $locale);
+
+            /** @var PageDocument $pageDocument */
+            $pageDocument = $this->documentManager->find($pageStructure->getUuid());
+
+            if ($pageStructure->getNodeType() === Structure::NODE_TYPE_EXTERNAL_LINK) {
+                $page['content']['url'] = $pageDocument->getRedirectExternal();
+            }
+
+            if ($pageStructure->getNodeType() === Structure::NODE_TYPE_INTERNAL_LINK) {
+                /** @var PageDocument $redirectTarget */
+                $redirectTarget = $pageDocument->getRedirectTarget();
+                $page['content']['url'] = $redirectTarget->getStructure()->getProperty('url')->getValue();
+            }
+
+            $pages[] = $page;
         }
 
-        return new ContentView($pages, ['ids' => $data ?: []]);
+        return new ContentView($pages, $data);
     }
 }
```

**Comment by alexander-schranz** [comment]
Thank you for report. We definitely should find a way doing this in the contentQueryBuilder at current state with your fix in sulu/sulu#5600 atleast the redirect should be done correctly.

**Comment by nnatter** [comment]
Yes, we should definitely provide some information about the target if a page is an internal link or external link. I am not sure if the `content` object is the correct place for this, because that object usually contains the resolved data of the selected page. I also think we should implement this somehow similar to the default Sulu implementation. 

In a clean sulu installation, the following property results in the following data:

```xml
<property name="pages" type="page_selection">
    <meta>
        <title>Pages</title>
    </meta>

    <params>
        <param name="properties" type="collection">
            <param name="title" value="title"/>
            <param name="url" value="url"/>
        </param>
    </params>
</property>
```

```php
"pages" => array:3 [▼
  0 => array:20 [▼
    "id" => "ff6e57bf-59e9-44c4-a8a0-e1e1565f3713"
    "uuid" => "ff6e57bf-59e9-44c4-a8a0-e1e1565f3713"
    "nodeType" => 1
    "path" => "/normal-page"
    "changed" => DateTime @1607515641 {#3674 ▶}
    "changer" => 1
    "created" => DateTime @1606314684 {#3677 ▶}
    "publishedState" => true
    "published" => DateTime @1606314684 {#3666 ▶}
    "creator" => 1
    "title" => "Normal Page"
    "locale" => "en"
    "webspaceKey" => "example"
    "template" => "default"
    "parent" => "09d5d72e-8e3c-44fe-bf50-fecfbc0146c6"
    "url" => "/normal-page"
    "urls" => array:1 [▼
      "en" => "/normal-page"
    ]
    "author" => "1"
    "authored" => DateTime @1606314683 {#3669 ▶}
    "order" => 20
  ]
  1 => array:20 [▼
    "id" => "8c9d007c-45dc-4094-8e4b-4d4c1579f443"
    "uuid" => "8c9d007c-45dc-4094-8e4b-4d4c1579f443"
    "nodeType" => 2
    "path" => "/internal-link"
    "changed" => DateTime @1607515806 {#4300 ▶}
    "changer" => 1
    "created" => DateTime @1606897107 {#4297 ▶}
    "publishedState" => true
    "published" => DateTime @1606897107 {#4309 ▶}
    "creator" => 1
    "title" => "Internal Link Target"
    "locale" => "en"
    "webspaceKey" => "example"
    "template" => "default"
    "parent" => "09d5d72e-8e3c-44fe-bf50-fecfbc0146c6"
    "url" => "/internal-link-target"
    "urls" => array:1 [▼
      "en" => "/internal-link-target"
    ]
    "author" => "1"
    "authored" => DateTime @1606897106 {#4292 ▶}
    "order" => 70
  ]
  2 => array:20 [▼
    "id" => "13cc6cfd-6d02-4565-9c92-7270b1d71093"
    "uuid" => "13cc6cfd-6d02-4565-9c92-7270b1d71093"
    "nodeType" => 4
    "path" => "/external-link"
    "changed" => DateTime @1606720220 {#4422 ▶}
    "changer" => 1
    "created" => DateTime @1606720205 {#4386 ▶}
    "publishedState" => true
    "published" => DateTime @1606720205 {#4413 ▶}
    "creator" => 1
    "title" => "External Link"
    "locale" => "en"
    "webspaceKey" => "example"
    "template" => "default"
    "parent" => "09d5d72e-8e3c-44fe-bf50-fecfbc0146c6"
    "url" => "/external-link"
    "urls" => array:1 [▼
      "en" => "/external-link"
    ]
    "author" => "1"
    "authored" => DateTime @1606720205 {#4381 ▶}
    "order" => 60
  ]
]
```

**Comment by niklasnatter** [comment]
Thanks for reporting! Should be fixed by #98
